include/seastar/http/common.hh

```diff
@@ -59,6 +62,20 @@ public:
         return path(key);
     }
 
+    sstring get_decoded_param(const sstring& key) const {
+        auto res = params.find(key);
+        if (res == params.end()) {
+            return "";
+        }
+        auto raw_path_param = res->second.substr(1);
```

Comment:
Why substr(1)?

Reply:
The path matcher adds arguments into the internal map of params with a ‘/‘ prefix, so I made `get_decoded_param` remove this from the return value to match the behaviour of operator[] above on line 55.

I expect that in most cases people use `params["param-key"]` and expect `params.get_decoded_param("param-key")` to return the same parameter value, decoded, without the `/` at the front. I am wondering though whether we should replace the first 5 lines of `get_decoded_param`:
```
auto res = params.find(key);
if (res == params.end()) {
    return "";
}
auto raw_path_param = res->second.substr(1);
```
with 
```auto raw_path_param = operator[](key);```
to match the error behaviour in the case that the key doesn't exist.

Reply:
The more I look at this get_decoded_param() the less I like it.

Why would someone want to _avoid_ using this method and grab req.params directly? Isn't it more correct for the param matcher to decode the paths while parsing and put req.params already parsed?

Reply:
> But how do we know it? :)

Whoever owns the client and server implementation pair knows it. :) Seastar's http router doesn't know it on its own. 

> The only problem will happen in case client sends a path that needs to be en/de-coded, the path arrives encoded, but server handler explicitly wants un-decoded path parameter. Am I missing something else?

Decoding more than necessary can also be a problem. Two examples I can think of are:
(a) Let's say the API endpoint is `example.com/labels/{some-custom-label}` where `some-custom-label` is a path parameter that we expect to be an arbitrary string defined by some end user. Let's also say that today the client doesn't path encode the parameter and the server just uses the raw `req.param` to get the label. We could have one of the custom labels be `DISCOUNT%20` that some end-user defined. Now, suppose we change seastar to decode path parameters by default. In that case, it silently changes the behaviour in the server as it will now consider this label as `DISCOUNT ` instead of `DISCOUNT%20` as the customer intended.
(b) Similarly, you could have the same API but in this case the client *does* path encode and the server also path decodes `req.param` because seastar had not path decoded `req.param` in the past. Now again, if the end-user gives the label `DISCOUNT%20` to the client, the client will path encode it and call the endpoint with `example.com/labels/DISCOUNT%2520`. If seastar now path decodes the parameter, it will set `req.param` with `DISCOUNT%20`. After that, the server will still do its path decoding leading to `DISCOUNT ` again.

So that is why I think it would be safer to keep the current behaviour of `req.param` and expose path decoded parameters as a separate `get_path_param` method.

Reply:
> Let's also say that today the client doesn't path encode the parameter ...

I think it can be treated as buggy client, and we can assume that paths coming from client are always encoded. I checked the `request::parse_query_param()` -- it assumes that query parameters are encoded and decodes them inside `add_query_param()`, but it keeps path as-is

> We could have one of the custom labels be DISCOUNT%20 that some end-user defined. Now, suppose we change seastar to decode path parameters by default. In that case, it silently changes the behaviour in the server as it will now consider this label as DISCOUNT instead of DISCOUNT%20 as the customer intended.

Wait, wait, the `DISCOUNT%20` is the as-is string passed by the client? In that case, it must be a "buggy client" right?

> (b) Similarly, you could have the same API but in this case the client does path encode and the server also path decodes req.param because seastar had not path decoded req.param in the past. Now again, if the end-user gives the label DISCOUNT%20 to the client, the client will path encode it and call the endpoint with example.com/labels/DISCOUNT%2520. If seastar now path decodes the parameter, it will set req.param with DISCOUNT%20. After that, the server will still do its path decoding leading to DISCOUNT again.

Yeah, double-decoding can be very surprising... In that case we need to think how to deprecate the old behavior. I mean:
- current API for creating/registering path_descriptions is marked as deprecated
- new way to create path appears
- when path is registered with new API, `parse_query_param()` always decodes it before processing

Reply:
I completely missed this PR... While rediscovering the missing path decoding, I also discovered the '+' bug - it appears somebody incorrectly added this HTML-only extension to the generic decoding function, which isn't a great idea.
I'm not working today so I can't review this PR today, but I'll do it on Sunday. Please review mine also.

Reply:
please also refer to the existing open Seastar issue.

Reply:
Namely, https://github.com/scylladb/seastar/issues/725

Reply:
I also don't understand the need for the new function or the talk about "double decoding". The old API for query parameters (NOT the path) worked, why do we need to change it? What am I missing?

Reply:
I just realized that operator[] also has that funky `substr(1)`. I still don't understand why it's done like this, but at least it's not some mistake - this is how it always was :-(